packages/test-utils/src/e2e.ts

```diff
@@ -0,0 +1,242 @@
+import assert from 'assert';
+import type { ChildProcess } from 'child_process';
+import { spawn } from 'child_process';
+import EventEmitter from 'events';
+import { join } from 'path';
+import stripAnsi from 'strip-ansi';
+import { fileURLToPath } from 'url';
+
+import { delay } from './delay.js';
+
+/**
+ * A test runner for running CLI commands. It keeps track of the stdout and
+ * stderr of the process, and emits events when new data is written to stdout or
+ * stderr.
+ */
+class CommandLineRunner extends EventEmitter {
+  readonly #process: ChildProcess;
+
+  readonly stdout: string[] = [];
+
+  readonly stderr: string[] = [];
+
+  readonly #defaultTimeout: number;
+
+  #output = '';
+
+  #exitCode?: number | null = null;
+
+  constructor(
+    command: string,
+    options: string[],
+    workingDirectory: string = process.cwd(),
+    defaultTimeout = 30000,
+  ) {
+    super();
+
+    this.#defaultTimeout = defaultTimeout;
+    this.#process = spawn(command, options, {
+      cwd: workingDirectory,
+      stdio: 'pipe',
+    });
+
+    this.#process.stdout?.on('data', (data) => {
+      const string = data.toString();
+      this.#output += string;
+
+      this.stdout.push(string);
+      this.emit('stdout', string);
+    });
+
+    this.#process.stderr?.on('data', (data) => {
+      const string = data.toString();
+      this.#output += string;
+
+      this.stderr.push(string);
+      this.emit('stderr', string);
+    });
+
+    this.#process.on('exit', (exitCode) => {
+      this.#exitCode = exitCode;
+      this.emit('exit', exitCode);
+    });
+  }
+
+  /**
+   * Whether the process is running.
+   *
+   * @returns `true` if the process is running, otherwise `false`.
+   */
+  get running() {
+    return this.#exitCode === null;
+  }
+
+  /**
+   * The output of the command, without ANSI escape codes.
+   *
+   * @returns The output of the command.
+   */
+  get output() {
+    return stripAnsi(this.#output);
+  }
+
+  /**
+   * Kill the process. If the process is already dead, this does nothing.
+   *
+   * @param signal - The signal to send to the process.
+   */
+  kill(signal?: NodeJS.Signals) {
+    if (!this.running) {
+      return;
+    }
+
+    this.#process.kill(signal);
+  }
+
+  /**
+   * Wait for the process to exit.
+   *
+   * @returns The exit code.
+   */
+  async wait() {
```

Comment:
This could be more explicit:
```suggestion
  async waitForExit() {
```